### PR TITLE
Ensure JIT/AOT code is not invalidated post-restore under `-XX:+DebugOnRestore`

### PR DESCRIPTION
**Pull request by dsouzai**
https://github.com/eclipse-openj9/openj9/pull/19754 adds VM Support for Debug On Restore. As the code stands, this will result in the JIT invalidating all code post-restore because it is unable to distinguish between FSD mode caused by the VM or the user.

This PR fixes this by adding the following changes:
* Cache the status of certain runtime events
* Return early from isFSDNeeded under `-XX:+DebugOnRestore`

Depends on https://github.com/eclipse-openj9/openj9/pull/19754 (specifically the update to the `isDebugOnRestoreEnabled` API).

### CODE REVIEW

**Comment by JasonFengJ9** [comment]
@dsouzai would you mind enabling the following failure conditions that were disabled in https://github.com/eclipse-openj9/openj9/pull/19754? I verified manually this PR passed the test with the failure conditions were reestablished. 
https://github.com/eclipse-openj9/openj9/blob/ebe8718d5838fc63d1ebeda5b1af37e2885ef320/test/functional/cmdLineTests/criu/criu_jitPostRestore.xml#L75-L77
https://github.com/eclipse-openj9/openj9/blob/ebe8718d5838fc63d1ebeda5b1af37e2885ef320/test/functional/cmdLineTests/criu/criu_jitPostRestore.xml#L94-L96

**Comment by dsouzai** [comment]
@ymanton could you please review/merge?

**Review by ymanton — APPROVED** [review]
LGTM.

**Comment by ymanton** [comment]
Jenkins test sanity.functional xlinux,plinux,zlinux jdk11,jdk17

File: runtime/compiler/runtime/CRRuntime.hpp [hunk 1]
```diff
@@ -69,6 +69,13 @@ class CRRuntime
 
    CRRuntime(J9JITConfig *jitConfig, TR::CompilationInfo *compInfo);
 
+   /**
+    * @brief Cache the status of JVMTI events such as exception throw/catch
+    *        as well as whether method trace and FSD were enabled
+    *        pre-checkpoint.
+    */
+   cacheEventsStatus();
```
Comment (keithc-ca) [inline runtime/compiler/runtime/CRRuntime.hpp:77]: This should have an explicit return type (i.e. `void`).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by dsouzai** [comment]
Jenkins test sanity.functional xlinux,plinux,zlinux jdk11,jdk17

**Comment by dsouzai** [comment]
The ppc64le test failure are all the `Test -Xjit:exclude={*}` without `-XX:+DebugOnRestore`:
```
[2024-09-21T00:23:24.079Z] Output from test:
[2024-09-21T00:23:24.079Z]  [OUT] start running script
[2024-09-21T00:23:24.079Z]  [OUT] export GLIBC_TUNABLES=glibc.cpu.hwcaps=-XSAVEC,-XSAVE,-AVX2,-ERMS,-AVX,-AVX_Fast_Unaligned_Load
[2024-09-21T00:23:24.079Z]  [OUT] export LD_BIND_NOT=on
[2024-09-21T00:23:24.079Z]  [OUT] /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64le_linux_Personal_testList_0/jdkbinary/j2sdk-image/bin/java -XX:+EnableCRIUSupport  -Xjit  -cp /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64le_linux_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criu.jar org.openj9.criu.OptionsFileTest JitOptionsTest -Xjit:exclude={*} 1
[2024-09-21T00:23:24.079Z]  [OUT] Pre-checkpoint
[2024-09-21T00:23:24.079Z]  [OUT] main: Sat Sep 21 00:23:18 UTC 2024, Performing CRIUSupport.checkpointJVM(), System.currentTimeMillis(): 1726878198957, System.nanoTime(): 1363144591107177
[2024-09-21T00:23:24.079Z]  [OUT] JVMJITM043W AOT load and compilation disabled post restore.
[2024-09-21T00:23:24.079Z]  [OUT] JVMJITM044W Some or all compiled code in the code cache invalidated post restore.
[2024-09-21T00:23:24.079Z]  [OUT] Post-checkpoint
[2024-09-21T00:23:24.079Z]  [OUT] initiate restore
[2024-09-21T00:23:24.079Z]  [OUT] Removed test output files
[2024-09-21T00:23:24.079Z]  [OUT] finished script
[2024-09-21T00:23:24.079Z]  [ERR] /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64le_linux_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh: line 41: 2543286 Killed                  $2 -XX:+EnableCRIUSupport $3 -cp "$1/criu.jar" $4 $5 $6 > testOutput 2>&1
[2024-09-21T00:23:24.079Z] >> Success condition was found: [Output match: Killed]
[2024-09-21T00:23:24.079Z] >> Required condition was found: [Output match: Pre-checkpoint]
[2024-09-21T00:23:24.079Z] >> Success condition was found: [Output match: Post-checkpoint]
[2024-09-21T00:23:24.079Z] >> Failure condition was not found: [Output match: CRIU is not enabled]
[2024-09-21T00:23:24.080Z] >> Failure condition was not found: [Output match: Operation not permitted]
[2024-09-21T00:23:24.080Z] >> Success condition was found: [Output match: Some or all compiled code in the code cache invalidated post restore.]
[2024-09-21T00:23:24.080Z] >> Failure condition was not found: [Output match: JIT compilation disabled post restore.]
[2024-09-21T00:23:24.080Z] >> Failure condition was found: [Output match: AOT load and compilation disabled post restore.]
[2024-09-21T00:23:24.080Z] >> Success condition was not found: [Output match: Thread pid mismatch]
[2024-09-21T00:23:24.080Z] >> Success condition was not found: [Output match: do not match expected]
[2024-09-21T00:23:24.080Z] >> Success condition was not found: [Output match: Unable to create a thread:]
[2024-09-21T00:23:24.080Z] >> Failure condition was not found: [Output match: User requested Java dump using]
```
Because `AOT load and compilation disabled post restore.` gets printed out before `Some or all compiled code in the code cache invalidated post restore.`, I believe this is due to AOT being disabled because of the SCC or similar, rather than because of something like FSD. Essentially, I believe it gets printed out here
https://github.com/eclipse-openj9/openj9/blob/0213b042b1f5d69ca9d625323d1b907bc6f7d2ee/runtime/compiler/control/OptionsPostRestore.cpp#L862
rather than here
https://github.com/eclipse-openj9/openj9/blob/0213b042b1f5d69ca9d625323d1b907bc6f7d2ee/runtime/compiler/control/OptionsPostRestore.cpp#L752

@ymanton I believe you've also seen this issue before right?

**Comment by dsouzai** [comment]
@ymanton could this be merged?

**Pull request merged**